pkg/option/resolver/resolver.go

```diff
@@ -0,0 +1,285 @@
+	first := true
+	for _, source := range sources {
+		c, descs, err := ReadConfigSource(ctx, client, nodeName, source)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read config source %s: %w", source.String(), err)
+		}
+		log.Infof("Got %d config pairs from source %+v", len(c), source.String())
+		if !first {
+			for k := range c {
+				if matchKeys != nil && !(matchKeys.Has(k) == allowIfMatch) {
+					log.Warnf("Source %s has key %s, but it cannot be overridden", source.String(), k)
+					delete(c, k)
+				}
+			}
+		}
+		first = false
+		if len(c) != 0 {
+			config = mergeConfig(source, config, c)
+			sourceDescriptions = append(sourceDescriptions, descs...)
+		}
+	}
+
+	config["config-sources"] = strings.Join(sourceDescriptions, ",")
+
+	return config, nil
+}
+
+func mergeConfig(source ConfigSource, lower, upper map[string]string) map[string]string {
+	out := make(map[string]string, len(lower))
+
+	for k, v := range lower {
+		out[k] = v
+	}
+
+	for k, v := range upper {
+		if _, set := out[k]; set {
+			log.Infof("Source %s overrides key %s", source, k)
+		}
+		out[k] = v
+	}
+
+	return out
+}
+
+// WriteConfigurations writes the key-value pairs in data to destDir. It writes it
+// like a Kubernetes config-map: It uses a double-layer symlink to allow for
+// atomic updates:
+// destDir/key -> ..data/key
+// ..data -> ..data_$time
+func WriteConfigurations(ctx context.Context, destDir string, data map[string]string) error {
+	dataDirName := fmt.Sprintf("..data_%d", time.Now().Unix())
+	err := os.MkdirAll(filepath.Join(destDir, dataDirName), 0777)
+	if err != nil {
+		return fmt.Errorf("failed to create data directory %s", filepath.Join(destDir, dataDirName))
+	}
+
+	for k, v := range data {
+		if strings.ContainsRune(k, os.PathSeparator) {
+			log.Errorf("Ignoring key with path separator: %v", k)
+			continue
+		}
+
+		dest := filepath.Join(destDir, dataDirName, k)
+		if err := os.WriteFile(dest, []byte(v), 0644); err != nil {
+			return fmt.Errorf("failed to write config key at %s: %w", dest, err)
+		}
+	}
+
+	_ = os.Remove(filepath.Join(destDir, "..data.tmp"))
+
+	// can't atomically update symlinks, so create a new one and rename
+	if err := os.Symlink(dataDirName, filepath.Join(destDir, "..data.tmp")); err != nil {
+		return fmt.Errorf("failed to write ..data.tmp symlink: %w", err)
+	}
+	if err := os.Rename(filepath.Join(destDir, "..data.tmp"), filepath.Join(destDir, "..data")); err != nil {
+		return fmt.Errorf("failed to move ..data symlink in to place: %w", err)
+	}
+
+	for k := range data {
+		if err := os.Symlink(filepath.Join("..data", k), filepath.Join(destDir, k)); err != nil && !os.IsExist(err) {
+			return fmt.Errorf("failed to write key %s symlink: %w", k, err)
+		}
+	}
+
+	return nil
+}
+
+func ReadConfigSource(ctx context.Context, client client.Clientset, nodeName string, source ConfigSource) (config map[string]string, descriptions []string, err error) {
+	log.Infof("Reading configuration from %s", source.String())
```

Comment:
nit: We typically try to avoid formatting variations on logging statements, because those break structured logging and log aggregation systems since the strings are dynamic. In this case it doesn't matter quite as much since the code will be used from a separate commandline utility, but typically we would instead define this explicitly such as:
```suggestion
	log.WithFields(logrus.Fields{
	    logfields.Source: source,
	}).Info("Reading option configuration")
```

(Same for all Infof(), Errorf(), etc log statements in this file)

Reply:
OK, I added structured logging. Good catch.

Reply:
There's still some cases where there's now both structured logging and formatting versions of the log calls. I think the right way forward is we just set up a linter for this though, then we can clean up any remaining cases in the codebase at that time.